File: SumoArenaSDK/Clients/python/sumo_arena_client.py
#! /usr/bin/python

# This program is a simple example of SumoArena coding contest
# client. It handles the socket and JSON layers.
#
# You can use this code to write your own SumoArena client.
#
# All you need is to put your AI code between this kind of tags
# below in the source code :
#
#    #--------------------------------------------------------------
#    # Your code begins here
#    #--------------------------------------------------------------
#
# and
#
#    #--------------------------------------------------------------
#    # Your code ends here.
#    #--------------------------------------------------------------
#
# You may also want to change this information before joining the
# contest:

CLIENT_NAME     = 'Python Dummy Client' # <-- Put your own client name here.
AVATAR_URL      = ''                    # <-- Put a picture URL (PNG), if any.

# This client does nothing more than going to the right as fast
# as he can. It is unlikely to be the most clever strategy to
# win the contest.
#
# This code accepts command line arguments:
#
#           sumo_arena_client.py <hostname>:<port>
#
# where
#
#           <host_name> is the game server hostname (default: localhost)
#           <port_name> is the game server port (default: 9090)
#
# Good luck.

import sys

import asyncore
import socket

import json


GAME_SERVER_NAME = 'localhost'
GAME_SERVER_PORT = 9090


class StartGameInfo(object):

    def __init__(self, raw_data):
        self.__dict__.update(raw_data)


class Sphere(object):
    
    def __init__(self, raw_data):
        self.__dict__.update(raw_data)


class PlaygroundInformation(object):

    def __init__(self, raw_data):

        self.arenaRadius = raw_data["arenaRadius"]
        self.players = [Sphere(data) for data in raw_data["players"]]


class EndGameInfo(object):

    def __init__(self, raw_data):
        self.__dict__.update(raw_data)


class Player(object):

    def __init__(self, client):
        
        self._client = client

    def on_prepare_information(self, startInfo):
        
        # Get all the available information.
        self.ownIndex           = startInfo.yourIndex
        self.playerCount        = startInfo.playerCount
        self.arenaInitialRadius = startInfo.arenaInitialRadius
        self.radius             = startInfo.sphereRadius
        self.maxVar             = startInfo.maxSpeedVariation
        self.currentRound       = startInfo.currentRound
        self.roundForVictory    = startInfo.roundForVictory

        print "Round is about to start."
        print "Now playing..."

    def on_play_request(self, playgroundInfo):
        
        if not playgroundInfo.players[self.ownIndex].inArena:
            # I'm out of the arena.
            return None

        result = self._choose_acceleration(playgroundInfo)

        # Check given rules.
        #
        # Assert()ions would be a little hard on us, isn't it?

        if not isinstance(result, tuple):
            print "*** Acceleration is supposed to be a (int, int) tuple."
        if not isinstance(result[0], int):
            print "*** Acceleration is supposed to be a (int, int) tuple."
        if not isinstance(result[0], int):
            print "*** Acceleration is supposed to be a (int, int) tuple."
        if not len(result) == 2:
            print "*** Acceleration is supposed to be a (int, int) tuple."

        sq_norm = result[0] * result[0] + result[1] * result[1] 
        if sq_norm > self.maxVar * self.maxVar:
            print "*** Acceleration is too big."

        return result

    def _choose_acceleration(self, playgroundInfo):
        # Your tough job starts here.
        #
        # Your are expected to return an (int, int)
        # tuple that gives (dVx, dVy) acceleration.

        # Here is the available information.
        arenaRadius = playgroundInfo.arenaRadius

        # This one is your sphere.
        #
        # Other sphere information is the same.
        myself      = playgroundInfo.players[self.ownIndex]

        # Position.
        x           = myself.x
        y           = myself.y

        # Velocity.
        vx          = myself.vx
        vy          = myself.vy

        # Other info.
        index       = myself.index      # Sphere index in players list.
        inArena     = myself.inArena    # Tell if game is over for this sphere.

        #--------------------------------------------------------------
        # Your code begins here.
        #--------------------------------------------------------------

        # Replace with more clever decision.
        wanted_acceleration = self.maxVar, 0

        #--------------------------------------------------------------
        # Your code ends here.
        #--------------------------------------------------------------

        return wanted_acceleration

    def on_round_finished(self, endInfo):
        
        # Get all the available information.
        currentRound     = endInfo.currentRound
        gameWinnerIndex  = endInfo.gameWinnerIndex
        roundWinnerIndex = endInfo.roundWinnerIndex

        # Did I win the round?
        if roundWinnerIndex == self.ownIndex:
            print "Hey, I just won this round."
        else:
            print "Well, I did not win this round."

        # Is the game over?
        if gameWinnerIndex < 0:
            print "I lost a round, but I dit not lose the game."
            return

        # If so, am I the king?
        if gameWinnerIndex == self.ownIndex:
            print "Hey, I also won this game, it seems."
        else:
            print "Kudos to the brilliant winner (not me, sadly)."


class GameClient(asyncore.dispatcher):

    def __init__(self, address, name, avatar_url=''):

        self._player = None

        # Protocol automaton data.
        self._callbacks = {
                'acknowledgeConnection':    self._on_connection_ack,
                'prepare':                  self._on_prepare,
                'play':                     self._on_play,
                'finishRound':              self._on_finish_round
            }
        self._actions = self._callbacks.keys()

        # Prepara connection information and send it ASAP.
        self._to_send = '{ "action": "connectPlayer", "parameters": {"name": "' + name + '", "avatar_url": "' + avatar_url + '"} }'
        self._received = ''

        # Connect to the server.
        asyncore.dispatcher.__init__(self)
        self.create_socket(socket.AF_INET, socket.SOCK_STREAM)
        self.connect(address)

    def handle_connect(self):
        # Now we are connected, we need a brain.
        self._player = Player(self)

    def writable(self):
        return len(self._to_send) > 0

    def handle_write(self):
        # Send waiting data.
        sent = self.send(self._to_send)
        self._to_send = self._to_send[sent:]

    def handle_read(self):

        # Read what is available
        data = self.recv(1024)
        if not data:
            return
        self._received += data

        # Parse what we received.
        valid, action, params = self._parse_server_request(self._received)
        if valid:
            self._callbacks[action](params)
        else:
            print action, self._received
        self._received = '' # FIXME Could have problem with partial requests.

    def handle_close(self):
        print "Connection closed."
        self.close()

    def _parse_server_request(self, data):

        # Check JSON validity.
        req = None
        try:
            req = json.loads(data)
        except ValueError, ve:
            return False, 'Invalid JSON data', None

        # Check acceptable server actions.
        action = req.get('action', None)
        if not action:
            return False, 'No action in request', None
        if action not in self._actions:
            return False, 'Requested action is unknown', None

        # Get parameters.
        parameters = req.get('parameters', None)

        return True, action, parameters

    def _on_connection_ack(self, params):
        print "I'm connected as '%s'" % params["yourName"]

    def _on_prepare(self, params):
        info = StartGameInfo(params)
        self._player.on_prepare_information(info)

    def _on_play(self, params):
        info = PlaygroundInformation(params)
        result = self._player.on_play_request(info)
        if result:
            data = {
                "action":     "updateVector",
                "parameters": {
                    "dVx": result[0],
                    "dVy": result[1]
                }
            }
            self._to_send = json.dumps(data)

    def _on_finish_round(self, params):
        info = EndGameInfo(params)
        self._player.on_round_finished(info)


def main(argv=None):
    name = "Python example client"
    if len(argv) > 1:
        name = argv[1]
    client = GameClient( (GAME_SERVER_NAME, GAME_SERVER_PORT) , name)
    asyncore.loop()


if __name__ == "__main__":
    sys.exit( main(sys.argv) )

File: SumoArenaSDK/Clients/C#/SumoArenaClient.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Collections.Generic;

#region JSON Data structures
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

public class RoundStartInfo
{
    public static RoundStartInfo CreateFromJsonString(String jsonString)
    {
        return JsonConvert.DeserializeObject<RoundStartInfo>(jsonString);
    }

    public int yourIndex { get; set; }
    public int playerCount { get; set; }
    public int sphereRadius { get; set; }
    public int maxSpeedVariation { get; set; }
    public int arenaInitialRadius { get; set; }
    public int currentRound { get; set; }
    public int roundForVictory { get; set; }
}

public class Sphere
{
    public int index { get; set; }
    public int x { get; set; }
    public int y { get; set; }
    public int vx { get; set; }
    public int vy { get; set; }
    public bool inArena { get; set; }
}

public class PlayingdInfo
{
    public static PlayingdInfo CreateFromJsonString(String jsonString)
    {
        return JsonConvert.DeserializeObject<PlayingdInfo>(jsonString);
    }

    public int arenaRadius { get; set; }
    public List<Sphere> players { get; set; }
}

public class RoundEndInfo
{
    public static RoundEndInfo CreateFromJsonString(String jsonString)
    {
        return JsonConvert.DeserializeObject<RoundEndInfo>(jsonString);
    }

    public int currentRound { get; set; }
    public int gameWinnerIndex { get; set; }
    public int roundWinnerIndex { get; set; }
}

#endregion

#region Event-based Player class

public class Player
{
    public class AccelerationVector
    {
        public int dVx;
        public int dVy;
    }
    
    RoundStartInfo roundInfo = null;

    public void OnRoundStart(RoundStartInfo info)
    {
        roundInfo = info;

        Console.WriteLine("A round is about to start.");
    }

    public AccelerationVector OnPlayRequest(PlayingdInfo info)
    {
        if (null == roundInfo || !info.players[roundInfo.yourIndex].inArena)
        {
            return null;
        }

        AccelerationVector result = ChooseAcceleration(info);

        // Quick check.
        if (result.dVx * result.dVx + result.dVy * result.dVy > roundInfo.maxSpeedVariation * roundInfo.maxSpeedVariation)
        {
            Console.WriteLine("*** Acceleration is too big.");
        }

        return result;
    }

    private AccelerationVector ChooseAcceleration(PlayingdInfo info)
    {
        // Get your own sphere.
        //
        // Other sphere information is the same.
        Sphere myself = info.players[roundInfo.yourIndex];

        // Position.
        int x = myself.x;
        int y = myself.y;

        // Velocity.
        int vx = myself.vx;
        int vy = myself.vy;

        // Other information.
        int index    = myself.index;
        bool inArena = myself.inArena;

        // --------------------------------------------------------------
        // Your code begins here.
        // --------------------------------------------------------------

        // You are expected to return an AccelerationVector.
        AccelerationVector result = new AccelerationVector();

        // Return dummy value.
        result.dVx = roundInfo.maxSpeedVariation;
        result.dVy = 0;

        // --------------------------------------------------------------
        // Your code ends here.
        // --------------------------------------------------------------

        return result;
    }

    public void OnRoundStop(RoundEndInfo info)
    {
        if (info.roundWinnerIndex == roundInfo.yourIndex)
        {
            Console.WriteLine("I won this round.");
        }
        else
        {
            Console.WriteLine("I did not win this round.");
        }

        if (info.gameWinnerIndex < 0)
        {
            Console.WriteLine("The game is not over yet.");
            return;
        }

        if (info.gameWinnerIndex == roundInfo.yourIndex)
        {
            Console.WriteLine("I won this game.");
        }
        else
        {
            Console.WriteLine("I did not win this game.");
        }
    }
}

#endregion
#region TCP Socket GameClient class

public class GameClient : TcpClient {

    private String clientName = String.Empty;
    private String avatarUrl  = String.Empty;

    private delegate void ActionMethod(String jsonParamString);
    private  Dictionary<String, ActionMethod> callbacks;

    private Player player;

    public GameClient(String hostname, UInt16 port, String name, String avatarUrl) : base(hostname, port)
    {
        this.clientName = name;
        this.avatarUrl  = avatarUrl;

        callbacks = new Dictionary<String, ActionMethod>() {
            {"acknowledgeConnection", new ActionMethod(OnConnectionAck)},
            {"prepare",               new ActionMethod(OnPrepare)},
            {"play",                  new ActionMethod(OnPlay)},
            {"finishRound",           new ActionMethod(OnFinishRound)}
        };
    }

    private void Run() {

        try
        {
            // Send identification message.
            if (Connected)
            {
                String jsonString = "{\"action\":\"connectPlayer\", \"parameters\": {\"name\":\"" + clientName + "\", \"avatar_url\": \"" + avatarUrl + "\" } }";
                byte[] jsonBytes = System.Text.Encoding.UTF8.GetBytes(jsonString);
                this.GetStream().Write(jsonBytes, 0, jsonBytes.Length);
            }
        
            // Now wait for server messages.
            while (Connected) {
                
                byte[] msgBytes = new byte[1024];
                int byteCount = this.GetStream().Read(msgBytes, 0, msgBytes.Length);

                if (byteCount < 0)
                {
                    continue;
                }

                String msg = System.Text.Encoding.UTF8.GetString(msgBytes);

                JObject o = JObject.Parse(msg);
                String action = o["action"].Value<String>();
                String parameters = o["parameters"].ToString();

                ActionMethod actionMethod = callbacks[action];
                actionMethod(parameters);
            }
            
        } catch (Exception e) {
            Console.WriteLine(e.ToString());
        }
    }

    private void OnConnectionAck(String jsonParamString)
    {
        JToken o = JObject.Parse(jsonParamString);
        Console.WriteLine("Now connected as {0}", o["yourName"].ToString());
        player = new Player();
    }

    private void OnPrepare(String jsonParamString)
    {
        Console.WriteLine("A round is starting.");
        RoundStartInfo info = RoundStartInfo.CreateFromJsonString(jsonParamString);
        player.OnRoundStart(info);
    }

    private void OnPlay(String jsonParamString)
    {
        PlayingdInfo info = PlayingdInfo.CreateFromJsonString(jsonParamString);
        Player.AccelerationVector v = player.OnPlayRequest(info);

        if (null != v)
        {
            String jsonString = String.Format("{{\"action\":\"updateVector\", \"parameters\": {{\"dVx\": {0}, \"dVy\": {1} }} }}", v.dVx, v.dVy);
            byte[] jsonBytes = System.Text.Encoding.UTF8.GetBytes(jsonString);
            this.GetStream().Write(jsonBytes, 0, jsonBytes.Length);
        }
    }

    private void OnFinishRound(String jsonParamString)
    {
        RoundEndInfo info = RoundEndInfo.CreateFromJsonString(jsonParamString);
        player.OnRoundStop(info);
    }

    public static int Main(String[] args) {

        // Default client argument values.
        const String DEFAULT_GAME_SERVER_HOST = "127.0.0.1";
        const UInt16 DEFAULT_GAME_SERVER_PORT = 9090;
        const String clientName = "C# Dummy client"; // <-- Put your client name here. 
	const String avatarUrl  = "";                // <-- Put a URL (PNG), if any.

        // Read information from command-line.
        String serverName = DEFAULT_GAME_SERVER_HOST;
        UInt16 serverPort = DEFAULT_GAME_SERVER_PORT;
        if (args.Length > 0)
        {
            Uri serverUri = new Uri(args[0]);
            serverPort = (UInt16) serverUri.Port;
            serverName = serverUri.Host;
        }

        GameClient gameClient = new GameClient(serverName, serverPort, clientName, avatarUrl);
        gameClient.Run();

        return 0;
    }
}

#endregion
